File: app/Services/SyncService.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Models\SyncLog;
use App\Models\SyncConflict;
use App\Models\Persona;
use App\Models\Evento;
use App\Models\Media;
use App\Models\Nota;
use App\Models\Tag;
use App\Models\PersonaLegame;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;
use Illuminate\Support\Facades\Storage;
use Carbon\Carbon;

class SyncService
{
    protected DiffService $diffService;

    public function __construct(DiffService $diffService)
    {
        $this->diffService = $diffService;
    }

    /**
     * Calcola e restituisce le differenze tra server e app
     */
    public function getDiff(array $appData, ?\DateTime $lastSyncTimestamp = null): array
    {
        return $this->diffService->calculateDiff($appData, $lastSyncTimestamp);
    }

    /**
     * Push: copia dati dall'app al server (App → Server)
     */
    public function pushFromApp(array $appData, string $syncMode = 'http', ?string $deviceId = null, array $mediaFiles = []): array
    {
        $syncLog = SyncLog::create([
            'sync_type' => 'push',
            'sync_mode' => $syncMode,
            'device_id' => $deviceId,
            'user_id' => auth()->check() ? auth()->id() : null,
            'status' => 'in_progress',
            'started_at' => now(),
        ]);

        try {
            $summary = [
                'persone' => 0,
                'eventi' => 0,
                'media' => 0,
                'note' => 0,
                'tags' => 0,
                'persona_legami' => 0,
            ];

            DB::transaction(function () use ($appData, &$summary, $mediaFiles) {
                // Sincronizza persone
                if (isset($appData['persone'])) {
                    $summary['persone'] = $this->syncTable('persone', Persona::class, $appData['persone'], false, []);
                }

                // Sincronizza eventi
                if (isset($appData['eventi'])) {
                    $summary['eventi'] = $this->syncTable('eventi', Evento::class, $appData['eventi'], false, []);
                }

                // Sincronizza media
                if (isset($appData['media'])) {
                    $summary['media'] = $this->syncTable('media', Media::class, $appData['media'], false, $mediaFiles);
                }

                // Sincronizza note
                if (isset($appData['note'])) {
                    $summary['note'] = $this->syncTable('note', Nota::class, $appData['note'], false, []);
                }

                // Sincronizza tags
                if (isset($appData['tags'])) {
                    $summary['tags'] = $this->syncTable('tags', Tag::class, $appData['tags'], false, []);
                }

                // Sincronizza persona_legami
                if (isset($appData['persona_legami'])) {
                    $summary['persona_legami'] = $this->syncTable('persona_legami', PersonaLegame::class, $appData['persona_legami'], false, []);
                }
            });

            $totalSynced = array_sum($summary);
            
            $syncLog->update([
                'status' => 'completed',
                'summary' => $summary,
                'records_synced' => $totalSynced,
                'completed_at' => now(),
            ]);

            return [
                'success' => true,
                'sync_log_id' => $syncLog->id,
                'summary' => $summary,
                'records_synced' => $totalSynced,
            ];
        } catch (\Exception $e) {
            Log::error('Sync push failed', ['error' => $e->getMessage(), 'trace' => $e->getTraceAsString()]);
            
            $syncLog->update([
                'status' => 'failed',
                'error_message' => $e->getMessage(),
                'completed_at' => now(),
            ]);

            throw $e;
        }
    }

    /**
     * Pull: invia dati dal server all'app (Server → App)
     */
    public function pullToApp(?\DateTime $lastSyncTimestamp = null): array
    {
        $data = [
            'persone' => $this->getTableData('persone', Persona::class, $lastSyncTimestamp),
            'eventi' => $this->getTableData('eventi', Evento::class, $lastSyncTimestamp),
            'media' => $this->getTableData('media', Media::class, $lastSyncTimestamp),
            'note' => $this->getTableData('note', Nota::class, $lastSyncTimestamp),
            'tags' => $this->getTableData('tags', Tag::class, $lastSyncTimestamp),
            'persona_legami' => $this->getTableData('persona_legami', PersonaLegame::class, $lastSyncTimestamp),
        ];

        $summary = [
            'persone' => count($data['persone']),
            'eventi' => count($data['eventi']),
            'media' => count($data['media']),
            'note' => count($data['note']),
            'tags' => count($data['tags']),
            'persona_legami' => count($data['persona_legami']),
        ];

        return [
            'data' => $data,
            'summary' => $summary,
            'total_records' => array_sum($summary),
            'sync_timestamp' => now()->toIso8601String(),
        ];
    }

    /**
     * Merge: unisce dati da entrambi i lati
     */
    public function merge(array $appData, ?\DateTime $lastSyncTimestamp = null, array $mediaFiles = []): array
    {
        $syncLog = SyncLog::create([
            'sync_type' => 'merge',
            'sync_mode' => 'http',
            'user_id' => auth()->check() ? auth()->id() : null,
            'status' => 'in_progress',
            'started_at' => now(),
        ]);

        // Inizializza variabili fuori dal try per sicurezza
        $summary = [
            'persone' => 0,
            'eventi' => 0,
            'media' => 0,
            'note' => 0,
            'tags' => 0,
            'persona_legami' => 0,
        ];
        $conflictsCount = 0;

        try {
            // Calcola differenze
            $diff = $this->diffService->calculateDiff($appData, $lastSyncTimestamp);
            
            // Reset summary per sicurezza
            $summary = [
                'persone' => 0,
                'eventi' => 0,
                'media' => 0,
                'note' => 0,
                'tags' => 0,
                'persona_legami' => 0,
            ];
            $conflictsCount = 0; // Inizializza fuori dalla closure
            
            DB::transaction(function () use ($appData, $diff, &$summary, &$conflictsCount, $mediaFiles) {
                // Applica modifiche: server → app (pull)
                foreach (['persone', 'eventi', 'media', 'note', 'tags', 'persona_legami'] as $table) {
                    if (!empty($diff[$table]['server_new'])) {
                        // Record nuovi sul server: aggiungi
                        $summary[$table] += count($diff[$table]['server_new']);
                    }
                    
                    if (!empty($diff[$table]['modified'])) {
                        // Record modificati: usa versione più recente
                        foreach ($diff[$table]['modified'] as $mod) {
                            if ($mod['type'] === 'server_newer') {
                                // Server più recente: aggiorna
                                $summary[$table]++;
                            }
                        }
                    }
                }

                // Applica modifiche: app → server (push)
                if (isset($appData['persone'])) {
                    $summary['persone'] += $this->syncTable('persone', Persona::class, $appData['persone'], true, []);
                }
                if (isset($appData['eventi'])) {
                    $summary['eventi'] += $this->syncTable('eventi', Evento::class, $appData['eventi'], true, []);
                }
                if (isset($appData['media'])) {
                    $summary['media'] += $this->syncTable('media', Media::class, $appData['media'], true, $mediaFiles);
                }
                if (isset($appData['note'])) {
                    $summary['note'] += $this->syncTable('note', Nota::class, $appData['note'], true, []);
                }
                if (isset($appData['tags'])) {
                    $summary['tags'] += $this->syncTable('tags', Tag::class, $appData['tags'], true, []);
                }
                if (isset($appData['persona_legami'])) {
                    $summary['persona_legami'] += $this->syncTable('persona_legami', PersonaLegame::class, $appData['persona_legami'], true, []);
                }

                // Gestisci conflitti
                foreach (['persone', 'eventi', 'media', 'note', 'tags', 'persona_legami'] as $table) {
                    if (!empty($diff[$table]['conflicts'])) {
                        foreach ($diff[$table]['conflicts'] as $conflict) {
                            // Verifica che i campi esistano prima di usarli
                            if (!isset($conflict['id']) || !isset($conflict['server_data']) || !isset($conflict['app_data'])) {
                                continue; // Salta conflitti malformati
                            }
                            
                            SyncConflict::create([
                                'table_name' => $table,
                                'record_id' => $conflict['id'],
                                'server_data' => $conflict['server_data'],
                                'app_data' => $conflict['app_data'],
                                'server_updated_at' => isset($conflict['server_updated_at']) 
                                    ? Carbon::parse($conflict['server_updated_at']) 
                                    : now(),
                                'app_updated_at' => isset($conflict['app_updated_at']) 
                                    ? Carbon::parse($conflict['app_updated_at']) 
                                    : now(),
                                'resolution' => 'pending',
                            ]);
                            $conflictsCount++;
                        }
                    }
                }
            });

            $totalSynced = array_sum($summary);
            
            $syncLog->update([
                'status' => 'completed',
                'summary' => $summary,
                'records_synced' => $totalSynced,
                'conflicts_found' => $conflictsCount,
                'completed_at' => now(),
            ]);

            return [
                'success' => true,
                'sync_log_id' => $syncLog->id,
                'summary' => $summary,
                'records_synced' => $totalSynced,
                'conflicts_found' => $conflictsCount,
            ];
        } catch (\Exception $e) {
            Log::error('Sync merge failed', [
                'error' => $e->getMessage(),
                'trace' => $e->getTraceAsString(),
                'file' => $e->getFile(),
                'line' => $e->getLine(),
            ]);
            
            $syncLog->update([
                'status' => 'failed',
                'error_message' => $e->getMessage(),
                'completed_at' => now(),
            ]);

            throw $e;
        }
    }

    /**
     * Sincronizza una tabella (upsert)
     */
    private function syncTable(string $tableName, string $modelClass, array $records, bool $mergeMode = false, array $mediaFiles = []): int
    {
        $count = 0;
        
        foreach ($records as $record) {
            // Rimuovi last_synced_at PRIMA di qualsiasi altra operazione
            // Questo evita problemi con formati di data non compatibili dall'app
            if (isset($record['last_synced_at'])) {
                unset($record['last_synced_at']);
            }
            
            // Rimuovi campi non fillable e id (gestito separatamente)
            $fillable = (new $modelClass)->getFillable();
            $id = $record['id'] ?? null;
            
            if (!$id) {
                continue; // Salta record senza ID
            }
            
            $data = array_intersect_key($record, array_flip($fillable));
            
            // TEMPORANEAMENTE DISABILITATO: last_synced_at causa problemi con SQLite
            // Aggiungi last_synced_at se presente nei fillable (sempre con now() per consistenza)
            // Non usare mai il valore dall'app, sempre generare uno nuovo
            // Converti in stringa per compatibilità con SQLite
            // if (in_array('last_synced_at', $fillable)) {
            //     $data['last_synced_at'] = now()->format('Y-m-d H:i:s');
            // }
            
            if ($mergeMode) {
                // In merge mode: usa versione più recente basata su updated_at
                $existing = $modelClass::find($id);
                if ($existing) {
                    // Converte Carbon a timestamp per confronto
                    $existingUpdated = $existing->updated_at 
                        ? ($existing->updated_at instanceof \Carbon\Carbon 
                            ? $existing->updated_at->timestamp 
                            : (is_string($existing->updated_at) 
                                ? strtotime($existing->updated_at) 
                                : 0)) 
                        : 0;
                    $newUpdated = isset($record['updated_at']) 
                        ? (is_string($record['updated_at']) 
                            ? strtotime($record['updated_at']) 
                            : ($record['updated_at'] instanceof \Carbon\Carbon
                                ? $record['updated_at']->timestamp
                                : (is_numeric($record['updated_at']) 
                                    ? (int)$record['updated_at'] 
                                    : 0))) 
                        : 0;
                    
                    if ($newUpdated > $existingUpdated) {
                        $existing->update($data);
                        $count++;
                        
                        // Se è un Media, verifica che il file esista e caricalo se necessario
                        if ($tableName === 'media' && $existing instanceof Media && isset($data['percorso'])) {
                            $this->handleMediaFile($existing, $data['percorso'], $mediaFiles);
                        }
                    }
                } else {
                    $data['id'] = $id; // Aggiungi ID per create
                    $created = $modelClass::create($data);
                    $count++;
                    
                    // Se è un Media, verifica che il file esista e caricalo se necessario
                    if ($tableName === 'media' && $created instanceof Media && isset($data['percorso'])) {
                        if ($this->handleMediaFile($created, $data['percorso'], $mediaFiles)) {
                            $filesUploaded++;
                        }
                    }
                }
            } else {
                // Push mode: sovrascrivi sempre
                $updated = $modelClass::updateOrCreate(['id' => $id], $data);
                $count++;
                
                // Se è un Media, verifica che il file esista e caricalo se necessario
                if ($tableName === 'media' && $updated instanceof Media && isset($data['percorso'])) {
                    if ($this->handleMediaFile($updated, $data['percorso'], $mediaFiles)) {
                        $filesUploaded++;
                    }
                }
            }
        }
        
        return $count;
    }
    
    /**
     * Gestisce il file media: verifica esistenza e carica se necessario
     * Restituisce true se il file è stato caricato, false altrimenti
     */
    private function handleMediaFile(Media $media, string $percorso, array $mediaFiles): bool
    {
        // Verifica se il file esiste già nel filesystem
        if (Storage::disk('public')->exists($percorso)) {
            return false; // File già presente, nessuna azione necessaria
        }

        // Cerca il file nei file caricati durante la sincronizzazione
        $fileName = basename($percorso);
        
        // Cerca il file per media_id (chiave come "media_id" o "media_files[media_id]")
        $uploadedFile = null;
        foreach ($mediaFiles as $key => $file) {
            // La chiave può essere "media_id" o "media_files[media_id]" dal multipart
            $keyStr = is_string($key) ? $key : (string)$key;
            if (strpos($keyStr, (string)$media->id) !== false || $keyStr === (string)$media->id) {
                $uploadedFile = $file;
                break;
            }
            // Fallback: cerca per nome file
            if ($file->getClientOriginalName() === $fileName || $file->getClientOriginalName() === $media->nome_file) {
                $uploadedFile = $file;
                break;
            }
        }

        if ($uploadedFile) {
            // Salva il file nel percorso specificato
            try {
                $directory = dirname($percorso);
                Storage::disk('public')->makeDirectory($directory);
                
                $savedPath = Storage::disk('public')->putFileAs(
                    $directory,
                    $uploadedFile,
                    $fileName
                );

                // Aggiorna dimensione e mime_type se necessario
                $media->update([
                    'dimensione' => $uploadedFile->getSize(),
                    'mime_type' => $uploadedFile->getMimeType(),
                ]);

                Log::info("Media file uploaded during sync", [
                    'media_id' => $media->id,
                    'percorso' => $savedPath,
                    'persona_id' => $media->persona_id,
                ]);
                
                return true; // File caricato con successo
            } catch (\Exception $e) {
                Log::error("Failed to upload media file during sync", [
                    'media_id' => $media->id,
                    'percorso' => $percorso,
                    'error' => $e->getMessage(),
                ]);
                return false;
            }
        } else {
            // File non trovato nei file caricati
            Log::warning("Media file missing during sync", [
                'media_id' => $media->id,
                'percorso' => $percorso,
                'persona_id' => $media->persona_id,
                'message' => 'Il file immagine non esiste nel filesystem e non è stato fornito durante la sincronizzazione.',
            ]);
            return false;
        }
    }

    /**
     * Ottiene dati di una tabella per il pull
     */
    private function getTableData(string $tableName, string $modelClass, ?\DateTime $lastSyncTimestamp): array
    {
        $query = $modelClass::query();
        
        if ($lastSyncTimestamp) {
            $query->where(function ($q) use ($lastSyncTimestamp) {
                $q->where('updated_at', '>', $lastSyncTimestamp)
                  ->orWhereNull('last_synced_at')
                  ->orWhere('last_synced_at', '<', $lastSyncTimestamp);
            });
        }

        $results = $query->get();
        
        // Converti in array e aggiungi last_synced_at se non presente
        return $results->map(function ($record) {
            $data = $record->toArray();
            if (!isset($data['last_synced_at'])) {
                $data['last_synced_at'] = null;
            }
            return $data;
        })->toArray();
    }
}

File: flutter_app/lib/services/sync_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:sqflite/sqflite.dart';
import '../core/database/database_helper.dart';
import 'api_client.dart';
import 'data_export_service.dart';
import 'adb_sync_service.dart';
import 'image_service.dart';
import '../data/local/persona_dao.dart';
import '../data/local/evento_dao.dart';
import '../data/local/media_dao.dart';
import '../data/local/nota_dao.dart';
import '../data/local/tag_dao.dart';
import '../data/local/persona_legame_dao.dart';

/// Servizio per la sincronizzazione bidirezionale con il server Laravel
class SyncService {
  final ApiClient _apiClient = ApiClient();
  final DataExportService _exportService = DataExportService();
  final AdbSyncService _adbSyncService = AdbSyncService();
  final ImageService _imageService = ImageService();
  final DatabaseHelper _dbHelper = DatabaseHelper.instance;
  final PersonaDao _personaDao = PersonaDao();
  final EventoDao _eventoDao = EventoDao();
  final MediaDao _mediaDao = MediaDao();
  final NotaDao _notaDao = NotaDao();
  final TagDao _tagDao = TagDao();
  final PersonaLegameDao _legameDao = PersonaLegameDao();

  /// Calcola le differenze tra database locale e server
  Future<Map<String, dynamic>> calculateDiff(DateTime? lastSyncTimestamp, {String syncMode = 'http'}) async {
    if (syncMode == 'adb') {
      // Per ADB, esporta i dati in un file e aspetta che lo script ADB li processi
      await _adbSyncService.exportDataToFile();
      throw Exception('Per ADB, usa lo script sync_via_adb.sh dal server. I dati sono stati esportati in sync_data.json');
    }

    // Modalità HTTP: calcola le differenze direttamente tramite API
    try {
      // Esporta i dati locali
      final appData = await _exportService.exportAllData();
      
      // Rimuovi il campo exported_at che non serve per il diff
      appData.remove('exported_at');

      // Chiama l'API per calcolare le differenze
      final response = await _apiClient.post('/sync/diff', {
        'app_data': appData,
        'last_sync_timestamp': lastSyncTimestamp?.toIso8601String(),
      });

      if (response['success'] == true) {
        return response['diff'] as Map<String, dynamic>;
      } else {
        throw Exception(response['message'] ?? 'Errore nel calcolo delle differenze');
      }
    } catch (e) {
      throw Exception('Errore durante il calcolo delle differenze: $e');
    }
  }

  /// Push: invia dati dall'app al server (App → Server)
  Future<Map<String, dynamic>> pushToServer(String syncMode) async {
    if (syncMode == 'adb') {
      // Per ADB, esporta i dati in un file e aspetta che lo script ADB li processi
      await _adbSyncService.exportDataToFile();
      throw Exception('Per ADB, esegui: ./sync_via_adb.sh push');
    }

    try {
      // Esporta i dati locali
      final appData = await _exportService.exportAllData();
      appData.remove('exported_at');

      // Invia al server
      final response = await _apiClient.post('/sync/push-from-app', {
        'app_data': appData,
        'sync_mode': syncMode,
      });

      if (response['success'] == true) {
        // TEMPORANEAMENTE DISABILITATO: last_synced_at causa problemi con SQLite
        // Aggiorna last_synced_at per tutti i record sincronizzati
        // await _updateLastSyncedAt();
        return response['data'] as Map<String, dynamic>;
      } else {
        throw Exception(response['message'] ?? 'Errore durante il push');
      }
    } catch (e) {
      throw Exception('Errore durante il push: $e');
    }
  }

  /// Pull: riceve dati dal server e li applica localmente (Server → App)
  Future<Map<String, dynamic>> pullFromServer(DateTime? lastSyncTimestamp, {String syncMode = 'http'}) async {
    if (syncMode == 'adb') {
      // Per ADB, verifica se ci sono dati da importare
      final hasData = await _adbSyncService.hasDataToImport();
      if (hasData) {
        await _adbSyncService.importDataFromFile();
        return {'success': true, 'message': 'Dati importati via ADB'};
      } else {
        throw Exception('Per ADB, esegui prima: ./sync_via_adb.sh pull');
      }
    }

    try {
      // Richiedi i dati dal server
      final response = await _apiClient.post('/sync/pull-to-app', {
        'last_sync_timestamp': lastSyncTimestamp?.toIso8601String(),
      });

      if (response['success'] == true) {
        final data = response['data'] as Map<String, dynamic>;
        
        // I dati sono direttamente in 'data', non in 'data.data'
        final serverData = data.containsKey('data') ? data['data'] as Map<String, dynamic> : data;

        // Applica i dati al database locale
        await _applyServerData(serverData);

        // TEMPORANEAMENTE DISABILITATO: last_synced_at causa problemi con SQLite
        // Aggiorna last_synced_at
        // await _updateLastSyncedAt();

        return data;
      } else {
        throw Exception(response['message'] ?? 'Errore durante il pull');
      }
    } catch (e) {
      throw Exception('Errore durante il pull: $e');
    }
  }

  /// Merge: unisce dati da entrambi i lati
  Future<Map<String, dynamic>> merge(DateTime? lastSyncTimestamp, String syncMode) async {
    if (syncMode == 'adb') {
      // Per ADB, esporta i dati e aspetta che lo script ADB faccia il merge
      await _adbSyncService.exportDataToFile();
      // Poi verifica se ci sono dati da importare dopo il merge
      final hasData = await _adbSyncService.hasDataToImport();
      if (hasData) {
        await _adbSyncService.importDataFromFile();
        return {'success': true, 'message': 'Merge completato via ADB'};
      } else {
        throw Exception('Per ADB, esegui: ./sync_via_adb.sh merge');
      }
    }

    try {
      // Esporta i dati locali
      final appData = await _exportService.exportAllData();
      appData.remove('exported_at');

      // Raccogli i file immagine da inviare
      final Map<String, File> mediaFiles = {};
      if (appData.containsKey('media')) {
        for (final media in appData['media'] as List) {
          final mediaId = media['id'];
          final percorso = media['percorso'] as String?;
          
          if (percorso != null && mediaId != null) {
            // Prova a trovare il file immagine locale
            final imageFile = await _imageService.getImageFile(percorso);
            if (imageFile != null && await imageFile.exists()) {
              // Usa media_id come chiave per identificare il file
              mediaFiles[mediaId.toString()] = imageFile;
            }
          }
        }
      }

      // Se ci sono file da inviare, usa multipart, altrimenti JSON normale
      final response = mediaFiles.isNotEmpty
          ? await _apiClient.postMultipart(
              '/sync/merge',
              {
                'app_data': appData,
                'last_sync_timestamp': lastSyncTimestamp?.toIso8601String(),
              },
              mediaFiles,
            )
          : await _apiClient.post('/sync/merge', {
              'app_data': appData,
              'last_sync_timestamp': lastSyncTimestamp?.toIso8601String(),
            });

      if (response['success'] == true) {
        final result = response['data'] as Map<String, dynamic>;
        
        // Se il server ha inviato dati aggiornati, applicali localmente
        if (result.containsKey('server_data')) {
          await _applyServerData(result['server_data'] as Map<String, dynamic>);
        }

        // TEMPORANEAMENTE DISABILITATO: last_synced_at causa problemi con SQLite
        // Aggiorna last_synced_at
        // await _updateLastSyncedAt();

        return result;
      } else {
        throw Exception(response['message'] ?? 'Errore durante il merge');
      }
    } catch (e) {
      throw Exception('Errore durante il merge: $e');
    }
  }

  /// Ottiene lo stato della sincronizzazione
  Future<Map<String, dynamic>> getStatus() async {
    try {
      final response = await _apiClient.get('/sync/status');
      if (response['success'] == true) {
        return response['data'] as Map<String, dynamic>;
      } else {
        throw Exception(response['message'] ?? 'Errore nel recupero dello stato');
      }
    } catch (e) {
      throw Exception('Errore durante il recupero dello stato: $e');
    }
  }

  /// Applica i dati ricevuti dal server al database locale
  Future<void> _applyServerData(Map<String, dynamic> serverData) async {
    final db = await _dbHelper.database;
    final now = DateTime.now().toIso8601String();

    await db.transaction((txn) async {
      // Applica persone
      if (serverData.containsKey('persone')) {
        for (final persona in serverData['persone'] as List) {
          final data = Map<String, dynamic>.from(persona);
          final id = data['id'];
          // TEMPORANEAMENTE DISABILITATO: last_synced_at causa problemi con SQLite
          // data['last_synced_at'] = now;
          if (id != null) {
            // Usa updateOrInsert per gestire sia nuovi che esistenti
            final count = await txn.update('persone', data, where: 'id = ?', whereArgs: [id]);
            if (count == 0) {
              await txn.insert('persone', data);
            }
          }
        }
      }

      // Applica eventi
      if (serverData.containsKey('eventi')) {
        for (final evento in serverData['eventi'] as List) {
          final data = Map<String, dynamic>.from(evento);
          final id = data['id'];
          // TEMPORANEAMENTE DISABILITATO: last_synced_at causa problemi con SQLite
          // data['last_synced_at'] = now;
          if (id != null) {
            final count = await txn.update('eventi', data, where: 'id = ?', whereArgs: [id]);
            if (count == 0) {
              await txn.insert('eventi', data);
            }
          }
        }
      }

      // Applica media
      if (serverData.containsKey('media')) {
        for (final media in serverData['media'] as List) {
          final data = Map<String, dynamic>.from(media);
          final id = data['id'];
          // TEMPORANEAMENTE DISABILITATO: last_synced_at causa problemi con SQLite
          // data['last_synced_at'] = now;
          if (id != null) {
            final count = await txn.update('media', data, where: 'id = ?', whereArgs: [id]);
            if (count == 0) {
              await txn.insert('media', data);
            }
          }
        }
      }

      // Applica note
      if (serverData.containsKey('note')) {
        for (final nota in serverData['note'] as List) {
          final data = Map<String, dynamic>.from(nota);
          final id = data['id'];
          // TEMPORANEAMENTE DISABILITATO: last_synced_at causa problemi con SQLite
          // data['last_synced_at'] = now;
          if (id != null) {
            final count = await txn.update('note', data, where: 'id = ?', whereArgs: [id]);
            if (count == 0) {
              await txn.insert('note', data);
            }
          }
        }
      }

      // Applica tags
      if (serverData.containsKey('tags')) {
        for (final tag in serverData['tags'] as List) {
          final data = Map<String, dynamic>.from(tag);
          final id = data['id'];
          // TEMPORANEAMENTE DISABILITATO: last_synced_at causa problemi con SQLite
          // data['last_synced_at'] = now;
          if (id != null) {
            final count = await txn.update('tags', data, where: 'id = ?', whereArgs: [id]);
            if (count == 0) {
              await txn.insert('tags', data);
            }
          }
        }
      }

      // Applica persona_legami
      if (serverData.containsKey('persona_legami')) {
        for (final legame in serverData['persona_legami'] as List) {
          final data = Map<String, dynamic>.from(legame);
          final id = data['id'];
          // TEMPORANEAMENTE DISABILITATO: last_synced_at causa problemi con SQLite
          // data['last_synced_at'] = now;
          if (id != null) {
            final count = await txn.update('persona_legami', data, where: 'id = ?', whereArgs: [id]);
            if (count == 0) {
              await txn.insert('persona_legami', data);
            }
          }
        }
      }
    });
  }

  /// Aggiorna last_synced_at per tutti i record
  Future<void> _updateLastSyncedAt() async {
    final db = await _dbHelper.database;
    final now = DateTime.now().toIso8601String();

    await db.transaction((txn) async {
      await txn.rawUpdate('UPDATE persone SET last_synced_at = ?', [now]);
      await txn.rawUpdate('UPDATE eventi SET last_synced_at = ?', [now]);
      await txn.rawUpdate('UPDATE media SET last_synced_at = ?', [now]);
      await txn.rawUpdate('UPDATE note SET last_synced_at = ?', [now]);
      await txn.rawUpdate('UPDATE tags SET last_synced_at = ?', [now]);
      await txn.rawUpdate('UPDATE persona_legami SET last_synced_at = ?', [now]);
    });
  }
}

